jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py

```diff
@@ -0,0 +1,136 @@
+                                              manual_options=['Mean', 'Min', 'Max', 'Sum'])  # noqa
+
+        self.add_results.viewer.filters = ['is_spectrum_viewer']
+
+    @property
+    def user_api(self):
+        return PluginUserApi(self, expose=('function', 'spatial_subset',
+                                           'add_results', 'collapse_to_spectrum'))
+
+    def collapse_to_spectrum(self, add_data=True, **kwargs):
+        """
+        Collapse over the spectral axis.
+
+        Parameters
+        ----------
+        add_data : bool
+            Whether to load the resulting data back into the application according to
+            ``add_results``.
+        kwargs : dict
+            Additional keyword arguments passed to the NDDataArray collapse operation.
+            Examples include ``propagate_uncertainties`` and ``ignore_masked_data``.
+        """
+        # get glue Data objects for the spectral cube and uncertainties
+        [spectral_cube] = self._app.get_data_from_viewer(
+            self._app._jdaviz_helper._default_flux_viewer_reference_name,
+            include_subsets=False
+        ).values()
+        [uncert_cube] = self._app.get_data_from_viewer(
+            self._app._jdaviz_helper._default_uncert_viewer_reference_name,
+            include_subsets=False
+        ).values()
+
+        # This plugin collapses over the *spatial axes* (optionally over a spatial subset,
+        # defaults to ``No Subset``). Since the Cubeviz parser puts the fluxes
+        # and uncertainties in different glue Data objects, we translate the spectral
+        # cube and its uncertainties into separate NDDataArrays, then combine them:
+        if self.spatial_subset_selected != self.spatial_subset.default_text:
+            nddata = spectral_cube.get_subset_object(
+                subset_id=self.spatial_subset_selected, cls=NDDataArray
+            )
+            uncertainties = uncert_cube.get_subset_object(
+                subset_id=self.spatial_subset_selected, cls=NDDataArray
+            ).data
+        else:
+            nddata = spectral_cube.get_object(cls=NDDataArray)
+            uncertainties = uncert_cube.get_object(cls=NDDataArray).data
+
+        # We attach the uncertainties to the NDDataArray with the data:
+        nddata.uncertainty = StdDevUncertainty(uncertainties)
+
+        # Collapse an e.g. 3D spectral cube to 1D spectrum, assuming that last axis
+        # is always wavelength. This will need adjustment after the following
+        # specutils PR is merged: https://github.com/astropy/specutils/pull/999
+        spatial_axes = tuple(range(nddata.ndim - 1))
+
+        # by default we want to use ignore_masked_data=True in nddata:
+        if "ignore_masked_data" not in kwargs:
+            kwargs["ignore_masked_data"] = True
+        # by default we want to propagate uncertainties:
+        if "propagate_uncertainties" not in kwargs:
+            kwargs["propagate_uncertainties"] = True
```

Comment:
would these be worth exposing as switches in the UI or are they meant to be advanced use-cases only for the API (in which case maybe they should be explicitly defined in the method, or are you expecting a bunch of other kwargs and need to keep it flexible)?


```suggestion
        # by default we want to use ignore_masked_data=True in nddata:
        kwargs.setdefault("ignore_masked_data", True)
        # by default we want to propagate uncertainties:
        kwargs.setdefault("propagate_uncertainties", True)
```

Reply:
This is a good question. The implications of the two options are a bit complex (see https://github.com/astropy/astropy/pull/14175#issuecomment-1440146039) so I'm torn between adding toggles for flexibility and leaving these subtleties in the astropy docs for more advanced users. Happy to hear all opinions!

Reply:
(un-auto-resolving so we can return to the conversation about these being explicitly listed as kwarg arguments and/or in the UI)